rmc/utils/constraint.py

```diff
@@ -1454,63 +1548,110 @@ def fix_region_start_stop_aas(
-    )
-    pos_ht = pos_ht.checkpoint(
-        f"{TEMP_PATH_WITH_FAST_DEL}/pos_per_transcript.ht",
+    missing_ht = missing_ht.checkpoint(
+        f"{TEMP_PATH_WITH_FAST_DEL}/region_start_stop_missing_aa.ht",
         _read_if_exists=not overwrite_temp,
         overwrite=overwrite_temp,
     )
-    missing_ht = missing_ht.annotate(
-        # Take the first missense position that is >= start coordinate
-        closest_start_mis=hl.bind(
-            lambda x: x[hl.binary_search(x, missing_ht.start_coordinate.position)],
-            pos_ht[missing_ht.transcript].positions,
-        ),
-        # Take the first missense position that is <= stop coordinate
-        closest_stop_mis=hl.bind(
-            lambda x: x[hl.binary_search(x, missing_ht.stop_coordinate.position) - 1],
-            pos_ht[missing_ht.transcript].positions,
-        ),
+    missing_transcripts = hl.literal(
+        missing_ht.aggregate(hl.agg.collect_as_set(missing_ht.transcript))
+    )
+
+    # Read in CDS HT resource
+    # Filter CDS and context HT to transcripts with internal missing start or stop AAs
+    context_ht = context_ht.filter(missing_transcripts.contains(context_ht.transcript))
+    cds_ht = transcript_cds.ht()
+    cds_ht = cds_ht.filter(missing_transcripts.contains(cds_ht.transcript))
+    cds_ht = cds_ht.annotate(
+        exon_start=cds_ht.interval.start,
+        exon_end=cds_ht.interval.end,
+    )
+
+    # Create two versions of CDS HT to fix missing region start and stop AAs
+    def _create_exon_position_ht(
+        cds_ht: hl.Table, fix_missing_start: bool, overwrite_temp: bool
+    ) -> hl.Table:
+        """
+        Create Table keyed by either exon start or stop position.
+
+        :param cds_ht: CDS HT resource filtered to transcripts missing region start or stop
+            AA annotations only.
+        :param fix_missing_start: Whether the output HT is designed to fix region starts
+            that are missing AA annotations.
+        :param overwrite_temp: Whether to overwrite temporary data.
+            If False, will read existing temp data rather than overwriting.
+            If True, will overwrite temp data.
+        :return: CDS HT keyed by either exon start or stop position and transcript.
+        """
+        if fix_missing_start:
+            # Region starts missing AAs need to get adjusted to use the AA from the previous
+            # exon start
+            # Reorder HT so that _prev_nonnull returns the next exon start
+            checkpoint_path = f"{TEMP_PATH_WITH_FAST_DEL}/cds_start_fix.ht"
+            cds_ht = cds_ht.order_by(
+                hl.asc(cds_ht.transcript), hl.desc(cds_ht.exon_start)
+            )
+            cds_ht = cds_ht.annotate(
+                next_exon_start=hl.scan._prev_nonnull(cds_ht.exon_start)
+            )
+            cds_ht = cds_ht.key_by(cds_ht.exon_stop, cds_ht.transcript)
+        else:
+            checkpoint_path = f"{TEMP_PATH_WITH_FAST_DEL}/cds_stop_fix.ht"
```

Comment:
I think a reorder here could possibly be needed just to ensure correctness since the original key is interval then transcript
```suggestion
            checkpoint_path = f"{TEMP_PATH_WITH_FAST_DEL}/cds_stop_fix.ht"
            cds_ht = cds_ht.order_by("transcript", "exon_stop")
```